crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -1,260 +1,535 @@
 use crate::{
     component::Component,
-    schedule::{Stage, SystemDescriptor, SystemStage},
-    world::World,
+    schedule::{ShouldRun, SystemSet},
+    system::{In, IntoChainSystem, IntoSystem, Local, Res, ResMut, System},
 };
-use bevy_utils::HashMap;
-use std::{mem::Discriminant, ops::Deref};
 use thiserror::Error;
 
-pub(crate) struct StateStages {
-    update: Box<dyn Stage>,
-    enter: Box<dyn Stage>,
-    exit: Box<dyn Stage>,
-}
-
-impl Default for StateStages {
-    fn default() -> Self {
-        Self {
-            enter: Box::new(SystemStage::parallel()),
-            update: Box::new(SystemStage::parallel()),
-            exit: Box::new(SystemStage::parallel()),
-        }
-    }
+/// ### Stack based state machine
+///
+/// This state machine has three operations: Next, Pop, and Push.
```

Comment:
```suggestion
/// This state machine has three operations: Push, Pop, and Next.
```
Following the order below.